npm-packages/cadence-parser/src/go.ts

```diff
@@ -0,0 +1,608 @@
+            return;
+          }
+          const toCopy = src.subarray(0, dst.length);
+          dst.set(toCopy);
+          setInt64(sp + 40, toCopy.length);
+          this.mem.setUint8(sp + 48, 1);
+        },
+
+        // func copyBytesToJS(dst ref, src []byte) (int, bool)
+        "syscall/js.copyBytesToJS": (sp) => {
+          sp >>>= 0;
+          const dst = loadValue(sp + 8);
+          const src = loadSlice(sp + 16);
+          if (!(dst instanceof Uint8Array || dst instanceof Uint8ClampedArray)) {
+            this.mem.setUint8(sp + 48, 0);
+            return;
+          }
+          const toCopy = src.subarray(0, dst.length);
+          dst.set(toCopy);
+          setInt64(sp + 40, toCopy.length);
+          this.mem.setUint8(sp + 48, 1);
+        },
+
+        "debug": (value) => {
+          console.log(value);
+        },
+      }
+    };
+  }
+
+  async run(instance) {
+    if (!(instance instanceof WebAssembly.Instance)) {
+      throw new Error("Go.run: WebAssembly.Instance expected");
+    }
+    this._inst = instance;
+    this.mem = new DataView(this._inst.exports.mem.buffer);
+    this._values = [ // JS values that Go currently has references to, indexed by reference id
+      NaN,
+      0,
+      null,
+      true,
+      false,
+      global,
+      this,
+    ];
+    this._goRefCounts = new Array(this._values.length).fill(Infinity); // number of references that Go has to a JS value, indexed by reference id
+    this._ids = new Map([ // mapping from JS values to reference ids
+      [0, 1],
+      [null, 2],
+      [true, 3],
+      [false, 4],
+      [global, 5],
+      [this, 6],
+    ] as [unknown, number][]);
+    this._idPool = [];   // unused ids that have been garbage collected
+    this.exited = false; // whether the Go program has exited
+
+    // Pass command line arguments and environment variables to WebAssembly by writing them to the linear memory.
+    let offset = 4096;
+
+    const strPtr = (str: string) => {
+      const ptr = offset;
+      const bytes = encoder.encode(str + "\0");
+      new Uint8Array(this.mem.buffer, offset, bytes.length).set(bytes);
+      offset += bytes.length;
+      if (offset % 8 !== 0) {
+        offset += 8 - (offset % 8);
+      }
+      return ptr;
+    };
+
+    const argc = this.argv.length;
+
+    const argvPtrs: number[] = [];
+    this.argv.forEach((arg) => {
+      argvPtrs.push(strPtr(arg));
+    });
+    argvPtrs.push(0);
+
+    const keys = Object.keys(this.env).sort();
+    keys.forEach((key) => {
+      argvPtrs.push(strPtr(`${key}=${this.env[key]}`));
+    });
+    argvPtrs.push(0);
+
+    const argv = offset;
+    argvPtrs.forEach((ptr) => {
+      this.mem.setUint32(offset, ptr, true);
+      this.mem.setUint32(offset + 4, 0, true);
+      offset += 8;
+    });
+
+    // The linker guarantees global data starts from at least wasmMinDataAddr.
+    // Keep in sync with cmd/link/internal/ld/data.go:wasmMinDataAddr.
+    const wasmMinDataAddr = 4096 + 4096;
```

Comment:
this is different in 1.18 https://github.com/golang/go/blob/1d9d99b7ce279f2af928f79cbc5906d99f29bb67/src/cmd/link/internal/ld/data.go#L2459-L2463

Maybe we should first bump `wasm_exec.js` to the latest and then convert it to typescript?

Reply:
Yes, I started trying to do this

Reply:
Good catch! Yes, unfortunately Go's WebAssembly runtime (`wasm_exec.js`) is version specific and we should bump it to 1.18. We can do this in a separate PR though